File: Client/src/main/java/com/rspsi/editor/model/LocalTile.java
package com.rspsi.editor.model;

/**
 * Tile coordinate local to one {@link WorldDocument}.
 *
 * <p>This type must never be used for OSRS absolute world coordinates. The
 * document boundary accepts LocalTile so an absolute viewport pick cannot be
 * passed into a 64x64 document accidentally.</p>
 */
public record LocalTile(int plane, int x, int y) {
    public LocalTile {
        if (plane < 0 || x < 0 || y < 0) {
            throw new IllegalArgumentException("Local tile coordinates cannot be negative");
        }
    }

    /** Compatibility bridge while older command/selection contracts migrate. */
    public TileCoordinate coordinate() {
        return new TileCoordinate(plane, x, y);
    }

    public static LocalTile from(TileCoordinate coordinate) {
        if (coordinate == null) throw new NullPointerException("coordinate");
        return new LocalTile(coordinate.plane(), coordinate.x(), coordinate.y());
    }
}

File: Client/src/main/kotlin/com/rspsi/editor/model/LocalTile.kt
package com.rspsi.editor.model

/**
 * Tile coordinate local to one [WorldDocument].
 *
 * This type must never represent absolute OSRS world coordinates. Keeping LocalTile and
 * WorldTile as distinct JVM-record types prevents document-local and world-space coordinates
 * from being mixed accidentally while preserving the Java record ABI during migration.
 */
@JvmRecord
data class LocalTile(
    val plane: Int,
    val x: Int,
    val y: Int,
) {
    init {
        if (plane < 0 || x < 0 || y < 0) {
            throw IllegalArgumentException("Local tile coordinates cannot be negative")
        }
    }

    /** Compatibility bridge while older command/selection contracts still use TileCoordinate. */
    fun coordinate(): TileCoordinate = TileCoordinate(plane, x, y)

    override fun toString(): String =
        "LocalTile[plane=$plane, x=$x, y=$y]"

    companion object {
        @JvmStatic
        fun from(coordinate: TileCoordinate?): LocalTile {
            val safeCoordinate = coordinate ?: throw NullPointerException("coordinate")
            return LocalTile(
                safeCoordinate.plane,
                safeCoordinate.x,
                safeCoordinate.y,
            )
        }
    }
}
